### DISCUSSION TITLE
Problem with NextJS 14.1.0 and Internationalization with 'use client' page or component

### DISCUSSION THREAD

**Discussion by james-devapps** [vercel/next.js#61107] [App Router] [answered]
Hi there,

I'm in a project and need two languages, coding an app following this this guide: https://nextjs.org/docs/app/building-your-application/routing/internationalization

Seems work all ok, but when I need use the translation in client side ('use client'), appears this error:

_You're importing a component that needs server-only. That only works in a Server Component which is not supported in the pages/ directory._

And I triyng make a component with 'use client', and appears the same error.

I try adapt this example but no have lucky (don't use TypeScript) https://github.com/vercel/next.js/tree/canary/examples/app-dir-i18n-routing

My code is here: https://github.com/james-devapps/nextjs-14-internationalization/tree/main

Any help?

Thanks in advance!!

**Comment by icyJoseph** [comment, marked as answer]
Hi,

Here, https://github.com/james-devapps/nextjs-14-internationalization/blob/main/src/languages.js, you are doing what's called, poisoning the module, so that it can only ever exist on the server.

When you import languages into a `use client` component, you tell the bundler to carry all of the imports, direct and indirect, into the client.

That's a contradiction.

What you can do is refactor your counter:

```js
'use client'

import { useState } from "react";

export default function Counter({ lang, actionLabel, title}) {
  const [count, setCount] = useState(0);
  
  return (
    <section>
      <p>
       {title}
      </p>
      <button onClick={() => setCount((n) => n + 1)}>
        {actionLabel}
      </button>
      {count}
    </section>
  );
}
```

Now the counter can take slots, which are just like children props, but with a meaningful name.

So you can use your counter from a server component, like your test-page as:

```jsx
import { getLanguages } from '@/languages'
import Counter from '../components/counter'

export default async function Home({ params: {lang} }) {
  const language = await getLanguages(lang)

  return (
    <main className="flex min-h-screen flex-col items-center justify-between p-24">
      <section className='my-12'>
        <p>Page with client component</p>
        <p>Current locale: {lang}</p>
        <p>{language.products.cart}</p>

        <Counter title="Counter" lang={lang} actionLabel={language.counter.increment} />
      </section>
    </main>
  )
}
```

React knows how to stitch together the server components and client components through this composition pattern.

**Reply by james-devapps** [reply]
Hi icyJoseph, work's perfect!

I had the concept wrong, thank you very much for the explanation and your time.

Do you see the rest of the code correct? Is there any bad practice?

Thank you!!
